File: 프로그래머스/3/43238. 입국심사/입국심사.java
class Solution {
    
    static boolean check(long ans, int[] times, int n){
        long temp =0;
        for(int i=0; i<times.length; i++){
            temp += (ans / (long)times[i]);
        }
        return temp >= n;
    }
    
    public long solution(int n, int[] times) {
        long answer = 0;
        long temp =1000000000;
        for (int i=0; i<times.length; i++){
            temp = Math.min(temp, times[i]);
        }
        long l = 1, r = temp * n;
        while(l <= r){
            long mid = (l + r) / 2;
            if(check(mid,times,n)){
                answer = mid;
                r = mid - 1;
            }
            else{
                l = mid + 1;
            }
        }
        
        
        return answer;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>

using namespace std;

bool check(long long mid, int n, vector<int> times){
    long long temp = 0;
    long long ans = (long long)n;
    for (int i= 0; i<times.size(); i++){
        temp += (mid / (long long)times[i]);
    }
    return temp >= ans;
}


long long solution(int n, vector<int> times) {
    long long answer = 0;
    long long l = 0, r = 900000000000000000;
    while(l<=r){
        long long mid = (l + r) / 2;
        if (check(mid, n, times)){
            answer = mid;
            r = mid-1;
        } else l = mid + 1;
    }
    
    return answer;
}